shared/subspace-metrics/src/lib.rs

```diff
@@ -0,0 +1,94 @@
+#![warn(missing_docs)]
+//! This Rust module serves as a bridge between two different Prometheus metrics libraries
+//! used within our frameworks — Substrate and Libp2p.
+//! The module exposes a web server endpoint at "/metrics" that outputs metrics in Prometheus
+//! format. It adapts metrics from either or both of the following libraries:
+//! - Official Rust Prometheus client (registry aliased as Libp2pMetricsRegistry)
+//! - TiKV's Prometheus client (registry aliased as SubstrateMetricsRegistry)
+
+
+use actix_web::http::StatusCode;
+use actix_web::web::Data;
+use actix_web::{get, App, HttpResponse, HttpServer};
+use parking_lot::Mutex;
+pub use prometheus::Registry as SubstrateMetricsRegistry;
+use prometheus::{Encoder, TextEncoder};
+use prometheus_client::encoding::text::encode;
+pub use prometheus_client::registry::Registry as Libp2pMetricsRegistry;
+use std::error::Error;
+use std::future::Future;
+use std::net::SocketAddr;
+use std::ops::DerefMut;
+use std::sync::Arc;
+use tracing::info;
+
+type SharedRegistry = Arc<Mutex<RegistryAdapter>>;
+
+/// An metrics registry adapter for Libp2p and Substrate frameworks.
+/// It specifies which metrics registry or registries are in use.
+pub enum RegistryAdapter {
+    /// Uses only the Libp2p metrics registry.
+    Libp2p(Libp2pMetricsRegistry),
+    /// Uses only the Substrate metrics registry.
+    Substrate(SubstrateMetricsRegistry),
+    /// We use both Substrate and Libp2p metrics registries.
+    Both(Libp2pMetricsRegistry, SubstrateMetricsRegistry),
+}
+
+#[get("/metrics")]
+async fn metrics(registry: Data<SharedRegistry>) -> Result<HttpResponse, Box<dyn Error>> {
+    let encoded_metrics = match registry.lock().deref_mut() {
```

Comment:
In this case, it would be possible to pass (None, None) that doesn't make sense. However, the code could be rewritten to have only two encoding snippets instead of four. Please, let me know if I need to simplify the code here.

Reply:
This is non-blocking, just not very nice API. It would be much nicer to have a builder API like this:
```rust
mod private {
    pub enum MetricsRegistry { Substrate(SubstrateRegistry), Official(Regitry) }

    impl From<SubstrateRegistry> for MetricsRegistry {}
    impl From<Registry> for MetricsRegistry {}
}

impl MetricsServer {
    pub fn new<Registry>(registry: Registry) -> Self
    where
        Registry: Into<private::MetricsRegistry> {}

    pub fn add_registry<Registry>(self, registry: Registry) -> Self
    where
        Registry: Into<private::MetricsRegistry> {}

    pub fn run(endpoints: Vec<SocketAddr>) -> std::io::Result<impl Future<Output = std::io::Result<()>>> {}
}
```

This way you can provide original registry as is in both cases without wrapping it into internal registry enum (that is not even exposed enymore).

Re-export could also be avoided, but yes, you'll have to include extra dependency, but I don't think that is a big deal.

Reply:
The proposed changes are more flexible (I would remove `add_registry` and change `new` parameter to vector). However, they will require parsing and sanitizing metrics (removing `#EOF` from libp2p versions at the very least). I would keep it as it is now until we have a more complex use case with multiple registries. 

I also added dependencies because I removed the reexport.

Reply:
I was talking about user-facing API, not flexibiluty or anything like that. It is awkward that you expose `SharedRegistry` and even more awkward that you call one of the metrics `Libp2p` even though it has literally nothing to do with lib2p as far as this crate is concerned, it is just misleading. All this can be hidden behind nice builder API.

And no, you can't use vector in `fn new` because then all the types must be the same and they will not be because we have different types of registries, this is why I have designed API in the way I did above.